Refactor add derivatives

pvandyken opened the pull request:
Here's an example of the reformatting I alluded to in #805. Most of the code was written just in my playing around with solutions: there's no tests yet, and just enough documentation to hopefully make things clear. So no worries if the idea isn't workable and must be discarded. Let me know if you have any thoughts.

# Issues
- With automatic derivative detection introduced in #805, derivative dataset validation became a necessary component of `BIDSLayout.__init__`. Previously, this was performed in `add_derivatives()`, and now any datasets added via `add_derivatives()` will be validated twice.
- There are now three separate locations where dataset `config` is set: within `BIDSLayout.__init__`, in `add_derivatives()`, and in `db.ConnectionManager.reset_database`

# Changes
Consolidation of the above jobs required a few important changes. The biggest issue was that, because dataset indexing is expensive, datasets should be fully validated before being indexed. This presents a problem for child datasets added via `add_derivatives()`, as they have a few additional validation features not present in a user-inited `BIDSLayout`:
  1. They must be derivatives: Currently the only way to check this is to wait until after indexing, or to read the `dataset_description.json` manually before calling `BIDSLayout.__init__`, largely negating the purpose of moving derivative validation fully into `__init__`.
  2. They must be indexed under a name (currently the Pipeline name), and this name cannot appear twice under the parent dataset (due to inherent limitations with `dict`s). This carries the same limitation as point 1: we must either wait until indexing is complete, or manually read `dataset_description.json` before calling `BIDSLayout`.


## Point 1
I added an `is_derivative` boolean parameter to `BIDSLayout.__init__` that, if True, would force the dataset to be indexed as a derivative. If `validate==True`, the dataset will be accordingly validated as a derivative, but no validation will occur otherwise (so even a dataset without a `dataset_description.json` could be treated as a derivative). Note that any derivative with a correctly formatted derivative `dataset_description.json` will still be indexed as a derivative even if `is_derivative==False`.

## Point 2
To address point 2: I changed the index name for child datasets from the *pipeline name* to the *directory name* of the dataset. For example, a dataset stored under `derivatives/fmriprep` would be indexed as `derivatives["fmriprep"]`, and a dataset stored under `foo/bar/some/pipeline` would be indexed as `derivatives["pipeline"]`. This solutions assumes that most users store their child datasets in a unique directory name that completely describes the data (users with a child dataset in a folder called `results` or similar would not find this change useful). One additional advantage that the change offers, however, is the possibility of indexing multiple versions of the same pipeline, so long as each version is stored under a different directory name (e.g. `fmriprep-1.8` vs. `fmriprep-1.9`). 

The behaviour of `BIDSLayout.get()` remains the same. All derivative datasets now have a `source_pipeline` attribute, and `get()` uses its value to find the correct scope.

The main breaking change occurs for users who attempt to directly query `BIDSLayout.derivatives[]`, as the keys of this `dict` have all been changed. As a shim, I implemented a `UserDict` that accepts both folder names and pipeline names as keys. Pipeline key access works through a method called `get_pipeline`. The current implementation of this method is intended to work with `__getitem__`, so it raises errors if either 0 or multiple items are found. However, this function could potentially be useful as a more powerful way of selecting child datasets (by pipeline, version, etc).

One other consequence of this change is that child database paths also use the directory name rather than the pipeline name. Here, this offers the additional advantage of ensuring a path safe name is being used for the database (since a Pipeline name can technically have any UTF). However, any existing saved databases would need to be reset. A shim could probably be built, but I don't have one yet.

## Misc

Finally, the three instances of setting `config` have all been consolidated into `BIDSLayout.__init__`.

effigies commented on bids/layout/models.py:
```diff
@@ -717,3 +723,45 @@ class FileAssociation(Base):
+
+        try:
+            result = self.get_pipeline(key)
+            print(
+                "[DEPRECATION WARNING] Directly selecting derivative datasets using "
+                "pipeline name (i.e. dataset.derivatives[<pipeline_name>] will be "
+                "phased out in an upcoming release. Select instead using the folder "
+                "name of the dataset (i.e. dataset.derivatives[<folder_name>]), or use "
+                "dataset.derivatives.get_pipeline(<pipeline_name>)."
+            )
```
This should be a deprecation warning.

```suggestion
            warnings.warn(
                "Directly selecting derivative datasets using pipeline name "
                "(i.e., dataset.derivatives[<pipeline_name>] will be phase out in "
                "an upcoming release. Select instead using the folder name of the "
                "dataset (i.e., dataset.derivatives[<folder_name>]), or use "
                "dataset.derivatives.get_pipeline(<pipeline_name>).",
                DeprecationWarning
            )
```

effigies commented on bids/layout/layout.py:
```diff
@@ -470,17 +508,33 @@ def add_derivatives(self, path, parent_database_path=None, **kwargs):
-        deriv_paths = validate_derivative_paths(paths, self, **kwargs)
+        deriv_paths = [*self._get_derivative_dirs(paths)]
+        if not deriv_paths:
+            warnings.warn("Derivative indexing was requested, but no valid "
+                            "datasets were found in the specified locations "
+                            "({}). Note that all BIDS-Derivatives datasets must"
+                            " meet all the requirements for BIDS-Raw datasets "
+                            "(a common problem is to fail to include a "
+                            "'dataset_description.json' file in derivatives "
+                            "datasets).\n".format(paths) +
+                            "Example contents of 'dataset_description.json':\n%s" %
+                            json.dumps(EXAMPLE_DERIVATIVES_DESCRIPTION))
```
```suggestion
            warnings.warn("Derivative indexing was requested, but no valid "
                          "datasets were found in the specified locations "
                          "({}). Note that all BIDS-Derivatives datasets must"
                          " meet all the requirements for BIDS-Raw datasets "
                          "(a common problem is to fail to include a "
                          "'dataset_description.json' file in derivatives "
                          "datasets).\n".format(paths) +
                          "Example contents of 'dataset_description.json':\n%s" %
                          json.dumps(EXAMPLE_DERIVATIVES_DESCRIPTION))
```

effigies commented on bids/layout/layout.py:
```diff
@@ -443,6 +463,24 @@ def parse_file_entities(self, filename, scope='all', entities=None,
 
+    def _get_derivative_dirs(self, paths):
+        def check_for_description(path):
+            return (path/"dataset_description.json").exists()
```
Personal preference to use `__class__.method(expr)` over `(expr).method()`.

```suggestion
            return Path.exists(path / "dataset_description.json")
```

effigies commented on bids/layout/db.py:
```diff
@@ -107,7 +107,6 @@ def reset_database(self, init_args, config=None):
             layout_info = LayoutInfo(**init_args)
         self.session.add(layout_info)
         # Add config records
-        config = listify('bids' if config is None else config)
```
If we don't want to use `['bids']` as a default, we should make `config` a required argument. Otherwise this is just going to cause the next line to fail if `reset_database()` is called without a config.

effigies reviewed (commented):
Thanks for your patience. Overall this looks good. Mostly nit-picky suggestions.

pvandyken commented:
Right, finally had a chance to go over this. I addressed all the comments above, and added a couple new tests. From my end, this seems to be everything necessary for merging, but let me know if there's anything else you'd like to address

pvandyken commented:
Actually, there's a few failure conditions in `DerivativeDatasets` that I didn't test. They're both fairly straightforward, but I'm happy to write tests when I get a chance (may not be till next week yet)

effigies commented:
To celebrate the anniversary of this PR, I wanted to check back in. IIRC it was mostly looking good, but your last comment was that you wanted to add more tests. Is that still the case?

pvandyken commented:
Yeah, I can do that. I'll try to get to it this weekend

pvandyken commented:
I have tests for all the missing coverage points, but it looks the ground has moved out from under me with #936 when it comes to validation. There's a particular conflict now in `add_derivatives()` where validation is explicitly skipped for derivatives. My code assumes derivatives CAN be validated in a sensible way. But now it looks like validate doesn't actually mean "enforce validation" any more, it means "only keep valid files". So I'm not really sure what to do with that.

I'll push my updates and let you decide. I'll have very limited time to understand the full implications of #936, but if you can suggest a specific way of responding I'll be happy to implement when I get a chance.

effigies commented:
> validate doesn't actually mean "enforce validation" any more, it means "only keep valid files"

That is most of what it's always meant. Though we also partially validated `dataset_description.json` .

Derivatives cannot currently be validated in a sensible way because the regular expressions in the `bids_validator` package are only for raw files. Constructing them for derivatives may be possible using the `bidsschematools` package, but we haven't gotten to the point of trying to use that. So for derivatives, it's really only been about validating the derivative 

> I'll push my updates and let you decide. I'll have very limited time to understand the full implications of #936, but if you can suggest a specific way of responding I'll be happy to implement when I get a chance.

The changes in #936 should not substantially affect this PR, but I see they've intertwined. I'll need some more time to analyze this one and make specific recommendations.

pvandyken commented:
> Derivatives cannot currently be validated in a sensible way because the regular expressions in the `bids_validator` package are only for raw files. Constructing them for derivatives may be possible using the `bidsschematools` package, but we haven't gotten to the point of trying to use that. So for derivatives, it's really only been about validating the derivative

Right, okay, so it would be better to say this PR introduces some *minimal* validation for derivatives. Basically just checking the dataset-description
